### DISCUSSION TITLE
Explicitly mark struct as unmanaged

### DISCUSSION THREAD

**Discussion by Tom01098** [dotnet/csharplang#2036] [Language Ideas] [answered]
This proposal is an extension of a small part of #688.

# Proposal
A struct should have the ability to be marked as unmanaged, so that analysis can happen pre-compile time. Much like the generic constraint, the `unmanaged` keyword in this context could imply the struct constraint, or like #688, it could just be an additional modifier.

```C#
public unmanaged UnmanagedStruct
{
    public int num;
}
```

In the event of a managed type being used at any level of nesting in the type, the compiler will emit an error.

```C#
public unmanaged AccidentallyManagedStruct
{
    public int num;
    public string whoops; // Not allowed!
}
```

`'AccidentallyManagedStruct' is marked as 'unmanaged', but 'whoops' is a managed type.`

# Motivation
It would prevent a developer from editing an unmanaged type, adding a string (or other managed type), and then trawling through other seemingly unrelated code.

# Drawbacks
Doesn't have many use cases, but would be especially useful for cases like the Unity Job System or low-level interop services.

**Comment by CyrusNajmabadi** [comment, marked as answer]
Can you not do this already today with an analyzer?  Just mark the structs however you want (an attribute would be fine), then just write an analyzer to check for these cases.

**Comment by Tom01098** [comment]
Yes, however I feel like this would make more sense to have wider language support rather than a Roslyn analyser as the keyword fits better with the generic constraint. The metadata that the attribute add isn't needed either as the type would just be a struct to the runtime.

**Comment by Zenexer** [comment]
I've started using analyzers for these sorts of restrictions.  There was an initial time investment to learn how they work, but they're quite effective and allow me to define all sorts of constraints like this one without much trouble.

That being said, I do see a benefit to this proposal: an `unmanaged struct` could satisfy an [`unmanaged` type parameter constraint](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.3/blittable).
